Backport section navigation improvements to Dogwood

pomegranited opened the pull request:
Backports the section navigation improvements merged to master for [PR 11705](https://github.com/edx/edx-platform/pull/11705), including subsequent bug fixes.

**JIRA tickets**: [OC-1574](https://tasks.opencraft.com/browse/OC-1574)

**Screenshots**:

Next button is enabled, even though we've navigated to the end of a section:
![prev_and_next](https://cloud.githubusercontent.com/assets/7556571/14911387/3216a34e-0e34-11e6-9636-b8a6bd909e65.png)

**Sandbox URL**: TBD - @bradenmacdonald should I build a sandbox for this?

**Testing instructions**:
1. Set up a devstack with OPENEDX_RELEASE="named-release/dogwood".
2. Checkout this repo/branch
3. Run LMS: `paver devstack lms`
4. Verify that you can use the courseware prev/next buttons to skip across sections.
5. Run tests: `paver test` 

**Author notes and concerns**:
1. I've tested that the section prev/next feature works in the LMS, but some (unrelated?) automated tests are still failing, probably because my cherry-picked commits aren't complete.  I will address those in the next couple of days.
2. I encountered [this issue with apt sources](https://www.reddit.com/r/chrome/comments/48oje6/linux_how_to_fix_failed_to_fetch/) when setting up a fresh devstack, which affected ansible's ability to install system packages.  The error can be seen by running `apt-get update` on your vagrant instance:
   
   ```
   Failed to fetch http://dl.google.com/linux/chrome/deb/dists/stable/Release
   Unable to find expected entry 'main/binary-i386/Packages' in Release file
   (Wrong sources.list entry or malformed file)
   ```
   
   Had to do both these steps to address it:
   
   ```
   sudo vim /etc/apt/sources.list.d/google-chrome.list  # correct line to include [arch=amd64]:
   # deb [arch=amd64] http://dl.google.com/linux/chrome/deb/ stable main
   sudo chattr +i /etc/apt/sources.list.d/google-chrome.list  # to prevent it from being overwritten
   ```

**Reviewers**
- [x] @jbzdak 
- [ ] other reviewers TBD

nasthagiri committed:
Activate Next and Previous Buttons across sections
MA-2152
MA-2153

(cherry picked from commit 8cbf99aca4274d51c57486a67b63015e3b488d2b)

nasthagiri committed:
Accessible Sequence navigation
MA-2190

(cherry picked from commit 2c90cac962183d2a1e23f4d585cfa75f4ab70544)

nasthagiri committed:
seq_module.html XSS - Safe by default

(cherry picked from commit 2f037f04e9be004f6fad7cb226ba254d0a2b1fb2)

Commit:
Update seq_* to edx.ui.lms.sequence.* format

seq_next, seq_prev, and seq_goto events are all renamed, and maintain
legacy compatibility.

This PR also introduces new EventTransformer framework to shim events
based on name prefix.

MA-2221

(cherry picked from commit 001874c42ed701fc221deb19d0fa16c7f63e179f)

nasthagiri committed:
Fix Next Button's 404 error

MA-2305

(cherry picked from commit 2396cc493d776a98918f467363ffbee994b35dff)

jillvogel committed:
Removes display_name_with_default_escaped

This method requires upstream commit which is out of scope for this change:

5e69224c32484ef9e0507f469babccd71a85c09b

jillvogel committed:
Removes gated content

Requires upstream commit
37a7fdc0f229b00c2e40fa387d86752603814c6a  (and others?)
which are not included in this change.

pomegranited commented on lms/envs/common.py:
```diff
@@ -397,6 +397,9 @@
     'ENABLE_LTI_PROVIDER': False,
+
+    # Enable Next Button to jump sequences in Sequence Navigation bar.
+    'ENABLE_NEXT_BUTTON_ACROSS_SECTIONS': False,
```
Note to self: This flag has [since been removed from master](https://github.com/edx/edx-platform/blame/master/lms/envs/common.py#L373), so should probably be removed from here too.

pomegranited commented on lms/envs/bok_choy.py:
```diff
@@ -157,6 +157,9 @@
 FEATURES['ENABLE_DASHBOARD_SEARCH'] = True
 
+# Enable cross-section Next button for tests
+FEATURES['ENABLE_NEXT_BUTTON_ACROSS_SECTIONS'] = True
```
Note to self: This flag has [since been removed from master](https://github.com/edx/edx-platform/blame/master/lms/envs/bok_choy.py#L160), so should probably be removed from here too.

bradenmacdonald commented on common/djangoapps/track/tests/test_shim.py:
```diff
@@ -149,3 +155,167 @@ def test_no_course_id(self):
+        }
+    }
+)
+class MultipleShimGoogleAnalyticsProcessorTestCase(EventTrackingTestCase):
```
Are all these event tracking changes required?

pomegranited replied:
No, sorry about that.. I only needed part of the commit, and have reverted the rest with [b8cdf17](https://github.com/hks-epod/edx-platform/pull/1/commits/b8cdf17e4e24b9bac1fa0326a49e73bd41dea108)

pomegranited committed:
Reverts most of cherry-picked commit 00187

Reverts the portions of 001874c42ed701fc221deb19d0fa16c7f63e179f not directly
related to the section sequencing issue.

pomegranited committed:
Removes a gated content test

Requires upstream commit
37a7fdc  (and others?)
which are not included in this change.

pomegranited committed:
Reverts unnecessary whitespace change

Changes to this file made by 8603a53c255d335ccad2d51d1a4f0cdb4d067241 were
largely reverted by d810e8665d6f1ebbca5b5c835896828ac740b8a1,
but left behind a whitespace change that we don't need.

pomegranited committed:
Reverts settings.FEATURES['ENABLE_NEXT_BUTTON_ACROSS_SECTIONS']

These settings were added by cherry-picked commit
8cbf99aca4274d51c57486a67b63015e3b488d2b, but not referenced anywhere, and so
were subsequently removed.

pomegranited committed:
Sets page_title = child.display_name_with_default

Makes the tests pass, while moving towards master:
https://github.com/edx/edx-platform/search?utf8=%E2%9C%93&q=tooltip_title

pomegranited committed:
Fixes Http404 errors thrown by courseware view tests

jbzdak commented:
@pomegranited Thanks for the google hint, it really helped! I had to alter steps you proposed (I'm posting this mostly to have these changes stored permanently somewehere). 
1. First I needed to add google signing key: `wget -q -O - https://dl.google.com/linux/linux_signing_key.pub | sudo apt-key add -`
2. Second due to some reason chrome sources file (`/etc/apt/sources.list.d/google-chrome`) got overwritten (despite `chattr`). So I just commented-out single entry in that file (commenting this entry out would is preserved by the google package scripts) and added a new file with proper sources list contents. 

I have tested your PR and it works. I'll be running automatic tests overnight.

bradenmacdonald commented:
@pomegranited Just noticed I didn't answer your question about sandboxes... No, no sandbox required, unless your reviewer (@jbzdak) says he needs one.

pomegranited commented:
Cool, thanks @bradenmacdonald .

pomegranited commented:
@jbzdak When I run the full test suite, I'm getting 3 errors from lms, which succeed when you run the tests by themselves:

```
python ./manage.py lms test --verbosity=1 lms/djangoapps/courseware/tests/test_submitting_problems.py  --settings=test
```

And 4 errors from the JS tests, which are reproducible, however there's an error in the js_test_tool that's preventing the errors from being properly reported.  I'm working on those errors now, don't know if I can fix the test tool though.

pomegranited commented:
@jbzdak @bradenmacdonald The tests are running ok for me now too, I just had to give the VM more memory to prevent timeouts.

Is there anything else we need to verify before merging?

jbzdak commented:
@pomegranited  I thought about requesting a sandbox, but then decided that this is the kind of change can be tested on devstack properly, and anyways the way we do deployment, we will have a chance to test right before we switch hks instance to new box, so there is no chance of pushing broken code to production. 

:+1:

pomegranited commented:
@bradenmacdonald I don't have write access to hks-epod/edx-platform.

If you're ok with this PR, could you please grant me access or do the merge?

jbzdak commented:
Merging on behalf of @pomegranited as she lacks permissions to do the merge herself.
